channel2/classic_listener.go

```diff
@@ -78,70 +83,98 @@ func (listener *classicListener) Create() (Underlay, error) {
+	logger := pfxlog.ContextLogger(listener.endpoint.String())
+	impl := newClassicImpl(peer, 2)
+	if connectionId, err := globalRegistry.newConnectionId(); err == nil {
+		impl.connectionId = connectionId
+		request, hello, err := listener.receiveHello(impl)
+		if err == nil {
+			for _, h := range listener.handlers {
+				logger.Infof("hello: %v, peer: %v, handler: %v", hello, peer, h)
+				if err := h.HandleConnection(hello, peer.PeerCertificates()); err != nil {
+					logger.Errorf("connection handler error (%s)", err)
+					if err := listener.ackHello(impl, request, false, err.Error()); err != nil {
+						logger.Errorf("error acknowledging hello (%s)", err)
 					}
-
-				} else {
-					log.Errorf("error receiving hello (%s)", err)
+					break
 				}
+			}
+
+			impl.id = &identity.TokenId{Token: hello.IdToken}
+			impl.headers = hello.Headers
+
+			if err := listener.ackHello(impl, request, true, ""); err == nil {
+				listener.created <- impl
 			} else {
-				log.Errorf("error getting connection id (%s)", err)
+				logger.Errorf("error acknowledging hello (%s)", err)
 			}
 
-		case <-listener.close:
-			return
+		} else {
+			logger.Errorf("error receiving hello (%s)", err)
 		}
+	} else {
+		logger.Errorf("error getting connection id (%s)", err)
 	}
+
+}
+
+type helloResponse struct {
+	message *Message
+	hello   *Hello
+	error   error
 }
 
 func (listener *classicListener) receiveHello(impl *classicImpl) (*Message, *Hello, error) {
-	log := pfxlog.ContextLogger(impl.Label())
-	log.Debug("started")
-	defer log.Debug("exited")
+	responseChan := make(chan helloResponse)
 
-	request, err := impl.rxHello()
-	if err != nil {
-		if err == UnknownVersionError {
-			writeUnknownVersionResponse(impl.peer.Writer())
+	go func() {
+		log := pfxlog.ContextLogger(impl.Label())
+		log.Debug("started")
+		defer log.Debug("exited")
+
+		request, err := impl.rxHello()
+		if err != nil {
+			if err == UnknownVersionError {
+				writeUnknownVersionResponse(impl.peer.Writer())
+			}
+			_ = impl.Close()
+			resp := helloResponse{error: fmt.Errorf("receive error (%s)", err)}
+			responseChan <- resp
 		}
-		_ = impl.Close()
-		return nil, nil, fmt.Errorf("receive error (%s)", err)
-	}
-	if request.ContentType != ContentTypeHelloType {
-		_ = impl.Close()
-		return nil, nil, fmt.Errorf("unexpected content type [%d]", request.ContentType)
+		if request.ContentType != ContentTypeHelloType {
+			_ = impl.Close()
+			resp := helloResponse{error: fmt.Errorf("unexpected content type [%d]", request.ContentType)}
+			responseChan <- resp
+		}
+		hello := UnmarshalHello(request)
+		responseChan <- helloResponse{message: request, hello: hello}
+	}()
+
+	select {
+	case helloResp := <-responseChan:
+
+		return helloResp.message, helloResp.hello, helloResp.error
+	case <-time.After(helloTimeout):
+		if err := impl.Close(); err != nil {
+			return nil, nil, fmt.Errorf("hello timed out, closing errored: %s", err)
```

Comment:
people can fill our disks and DoS us that way with this as an error... i like the thought but maybe it's better as debug?

Reply:
in fact that's true for all these log messages ... maybe not something we address now but in a new ticket?

Reply:
gah i keep forgetting - we leave log rolling to the operator - nevermind... it's on them to roll out the logs as appropriate i suppose